keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -228,14 +229,10 @@ class ResourceActuator(
           val artifact = deliveryConfig.matchingArtifactByName(versionedArtifact.artifactName, versionedArtifact.artifactType)
             ?: error("Artifact ${versionedArtifact.artifactType}:${versionedArtifact.artifactName} not found in delivery config ${deliveryConfig.name}")
 
-          val previousSuccess = artifactRepository.wasSuccessfullyDeployedTo(
-            deliveryConfig = deliveryConfig,
-            artifact = artifact,
-            version = versionedArtifact.artifactVersion,
-            targetEnvironment = environment.name
-          )
+          val promotionStatus = artifactRepository.getArtifactPromotionStatus(
+            deliveryConfig, artifact, versionedArtifact.artifactVersion, environment.name)
 
-          if (!previousSuccess) {
+          if (promotionStatus == DEPLOYING) {
```

Comment:
This was the problem... `wasSuccessfullyDeployedTo` checks for the presence of a record with a `CURRENT` or `PREVIOUS` promotion status, neither of which is the case with a just-approved version.

Reply:
@luispollo we also don't want to reject an artifact that has previously been deployed, so you'll also need to add in the CURRENT and PREVIOUS statuses

Reply:
The reason for this is that we don't want to mark something as bad that has been successfully deployed to an environment. In that case the artifact version is most likely totally fine. If we get to the point where the deployment fails of a version that was previously successful, we know it's probably an infra or env thing.

However, now that I'm thinking about it, is your point that artifacts go into the approved then deploying state always, and that we will never actually hit the case that the artifact is in a CURRENT or a PREVIOUS status? If so, I see what you're saying. That just means I think that we will have to track whether a version is successfully deployed in another place....

Reply:
> @luispollo we also don't want to reject an artifact that has previously been deployed, so you'll also need to add in the CURRENT and PREVIOUS statuses

Right. But this is the check for marking the artifact as bad, so we'll _not_ mark it as bad if it's `CURRENT` or `PREVIOUS` (or any of the other statuses, for that matter), only when it's `DEPLOYING`.



> The reason for this is that we don't want to mark something as bad that has been successfully deployed to an environment. In that case the artifact version is most likely totally fine. If we get to the point where the deployment fails of a version that was previously successful, we know it's probably an infra or env thing.
> 
> However, now that I'm thinking about it, is your point that artifacts go into the approved then deploying state always, and that we will never actually hit the case that the artifact is in a CURRENT or a PREVIOUS status? If so, I see what you're saying. That just means I think that we will have to track whether a version is successfully deployed in another place....

Yes, we don't currently have any historical tracking of artifact status transitions (we should probably introduce artifact events at some point). I tried to think of the smallest change I could make to achieve a close-enough behavior to what you were attempting to model before, because I think the real solution here would be to take a step back and revisit the way we're modeling this, as I've brought up before, but that would take much longer. 

With my change, if you get to the resource check and the latest artifact version is not the one that's currently deploying, then we don't touch the artifact. Which includes if the artifact had been previously deployed successfully (`PREVIOUS`) or is currently deployed successfully (`CURRENT`) and for some other reason the cluster is unhappy.

Reply:
👍 sounds good, I definitely read through this a bit too quickly so turns out my concerns were unfounded! 

I'll make a note to think about storing whether a version has been successfully deployed or not